### PR TITLE
MDS Support for trace analytics

### PR DESCRIPTION
**Pull request by sumukhswamy**
### Description
Added MDS Support for trace analytics

### Issues Resolved
[#1440 [FEATURE] Support Multiple Data Source in Observability Dashboards Plugin](https://github.com/opensearch-project/dashboards-observability/issues/1440)

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/trace_analytics/components/traces/trace_view.tsx [hunk 1]
```diff
@@ -51,9 +57,11 @@ export function TraceView(props: TraceViewProps) {
       </>
     );
   };
+  const DataSourceMenu = props.dataSourceManagement?.ui?.getDataSourceMenu<DataSourceViewConfig>();
 
   const renderOverview = (fields: any) => {
     return (
+      <>
```
Comment (joshuali925) [inline public/components/trace_analytics/components/traces/trace_view.tsx:64]: unnecessary change?

File: server/routes/trace_analytics_dsl_router.ts [hunk 2]
```diff
@@ -11,28 +11,38 @@ import {
...
+      const {dataSourceMDSId}  = request.query;
       try {
-        const resp = await context.core.opensearch.legacy.client.callAsCurrentUser(
-          'indices.exists',
-          params
-        );
+        let client;
+        let resp;
+        if (dataSourceEnabled && dataSourceMDSId) {
+          client = context.dataSource.opensearch.legacy.getClient(dataSourceMDSId);
```
Comment (joshuali925) [inline server/routes/trace_analytics_dsl_router.ts:41]: nit
```suggestion
        let resp;
        if (dataSourceEnabled && dataSourceMDSId) {
          const client = context.dataSource.opensearch.legacy.getClient(dataSourceMDSId);
```

File: public/components/trace_analytics/components/traces/traces_content.tsx [hunk 3]
```diff
@@ -78,12 +83,31 @@ export function TracesContent(props: TracesProps) {
...
+      {props.dataSourceEnabled && (
+        <DataSourceMenu
+          setMenuMountPoint={props.setActionMenu}
+          componentType={'DataSourceView'}
+          componentConfig={{
+            activeOption: dataSourceMDSId,
+            fullWidth: true,
+          }}
+        />
+      )}
```
Comment (joshuali925) [inline public/components/trace_analytics/components/traces/traces_content.tsx:110]: there was a DataSourceMenu in home.tsx, why is it needed again here?

File: public/components/trace_analytics/components/services/services_content.tsx [hunk 4]
```diff
@@ -54,6 +55,7 @@ export function ServicesContent(props: ServicesProps) {
   const [redirect, setRedirect] = useState(true);
   const [loading, setLoading] = useState(false);
   const [filteredService, setFilteredService] = useState('');
+  const DataSourceMenu = dataSourceManagement?.ui?.getDataSourceMenu<DataSourceViewConfig>();
```
Comment (joshuali925) [inline public/components/trace_analytics/components/services/services_content.tsx:58]: unused?

File: server/routes/trace_analytics_dsl_router.ts [hunk 5]
```diff
@@ -48,18 +58,28 @@ export function registerTraceAnalyticsDslRouter(router: IRouter) {
...
+          resp = await client.callAPI('indices.exists', params);
+        } else {
+          resp = await context.core.opensearch.legacy.client.callAsCurrentUser('indices.exists',params)
+        }
```
Comment (ykmr1224) [inline server/routes/trace_analytics_dsl_router.ts:82]: I see same pattern 3 times. Can we extract it as a function?

Plus, `client` can be const inside if clause.

File: server/routes/trace_analytics_dsl_router.ts [hunk 6]
```diff
@@ -11,28 +11,38 @@ import {
...
+          dataSourceMDSId: schema.maybe(schema.string({ defaultValue: '' }))
+        })
+      },
     },
     async (context, request, response) => {
       const params: RequestParams.IndicesExists = {
         index: [DATA_PREPPER_INDEX_NAME, DATA_PREPPER_SERVICE_INDEX_NAME],
         allow_no_indices: false,
       };
+      const {dataSourceMDSId}  = request.query;
```
Comment (ykmr1224) [inline server/routes/trace_analytics_dsl_router.ts:36]: nitnit: two spaces?

**Comment by YANG-DB** [comment]
Hi @sumukhswamy 
can you please explain in more details the overall change you've introduced here
 - flow changes
 - query updates
 - UX changes
Thanks

File: public/components/trace_analytics/components/traces/traces_content.tsx (hunk 3 above)
Comment (sumukhswamy) [reply public/components/trace_analytics/components/traces/traces_content.tsx:110]: there are two menus one is readonly and one is a selector

**Comment by sumukhswamy** [comment]
https://github.com/opensearch-project/dashboards-observability/assets/107440906/58b6050a-235f-4e28-a636-4e1e87426280

File: public/components/trace_analytics/components/services/service_view.tsx [hunk 7]
```diff
@@ -54,9 +62,23 @@ export function ServiceView(props: ServiceViewProps) {
...
       processTimeStamp(props.endTime, mode)
     );
-    handleServiceViewRequest(props.serviceName, props.http, DSL, setFields, mode);
+    handleServiceViewRequest(
+      props.serviceName,
+      props.http,
+      DSL,
+      setFields,
+      mode,
+      props.dataSourceMDSId[0].id
```
[the remark this answers is not in the document]
Comment (sumukhswamy) [reply public/components/trace_analytics/components/services/service_view.tsx:71]: have initialized it to [{id:'',label:''}]

**Review by joshuali925 — APPROVED** [review]

**Review by ps48 — APPROVED** [review]
